pkg/tls/certificatemanagement/ca.go

```diff
@@ -0,0 +1,596 @@
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/tls"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	CASecretName                = "tigera-ca-private"
+	TrustedCertConfigMapName    = "tigera-ca-bundle"
+	TrustedCertConfigMapKeyName = "tigera-ca-bundle.crt"
+	TrustedCertVolumeMountPath  = "/etc/pki/tls/certs/"
+	TrustedCertBundleMountPath  = "/etc/pki/tls/certs/tigera-ca-bundle.crt"
+)
+
+// CertificateManager can sign new certificates and has methods to retrieve existing KeyPairs and Certificates. If a user
+// brings their own secrets, CertificateManager will preserve and return them.
+type CertificateManager interface {
+	// CertificateManagement returns the CertificateManagement object or nil if it is not configured.
+	CertificateManagement() *operatorv1.CertificateManagement
+	// GetKeyPair returns an existing KeyPair. If the KeyPair is not found, nil is returned.
+	GetKeyPair(cli client.Client, secretName, secretNamespace string) (KeyPair, error)
+	// GetOrCreateKeyPair returns a KeyPair. If one exists, some checks are performed. Otherwise, a new KeyPair is created.
+	GetOrCreateKeyPair(cli client.Client, secretName, secretNamespace string, dnsNames []string) (KeyPair, error)
+	// GetCertificate returns a Certificate. If the certificate is not found a k8s.io NotFound error is returned.
+	GetCertificate(cli client.Client, secretName, secretNamespace string) (Certificate, error)
+	// Issued returns true if the provided certificate was signed by CertificateManager and returns false for user provided KeyPairs or Certificate Management.
+	Issued(certificate Certificate) bool
+	AddToStatusManager(manager status.StatusManager, namespace string)
+	KeyPair
+}
+
+// KeyPair wraps a Secret object that contains a private key and a certificate. Whether CertificateManagement is
+// configured or not, KeyPair returns the right InitContainer, VolumeMount or Volume (when applicable).
+type KeyPair interface {
+	//UseCertificateManagement returns true if this key pair was not user provided and certificate management has been configured.
+	UseCertificateManagement() bool
+	// BYO returns true if this KeyPair was provided by the user. If BYO is true, UseCertificateManagement is false.
+	BYO() bool
+	InitContainer(namespace string) corev1.Container
+	VolumeMount() corev1.VolumeMount
+	VolumeMountKeyFilePath() string
+	VolumeMountCertificateFilePath() string
+	Volume() corev1.Volume
+	Certificate
+}
+
+// Certificate wraps the certificate. Combine this with a TrustedBundle, to mount a trusted certificate bundle to a pod.
+type Certificate interface {
+	X509Certificate() *x509.Certificate
+	Secret(namespace string) *corev1.Secret
```

Comment:
Is this function ever actually used outside of this function on a Certificate? It's looking like this is actually a function of the KeyPair, not the Certificate.

Reply:
Looking at how this is used in this file I think some rethinking needs to happen here, you should need to convert this to a secret to get certificate / key values, and I don't think there's a good reason to use it at all in this file.

Any functions that create k8s objects in this file should be _**solely**_ for external purposes (i.e. you need to create a Secret from the cert / key / ca in this object).